pkg/common/scan.go

```diff
@@ -0,0 +1,28 @@
+package common
+
+import (
+	"bytes"
+	"regexp"
+)
+
+var newlineReg = regexp.MustCompile(`\r?\n|\r\n?`)
+
+func ScanLines(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	if atEOF && len(data) == 0 {
+		return 0, nil, nil
+	}
+	if loc := newlineReg.FindIndex(data); len(loc) > 0 {
```

Comment:
Looks like we don't need to use regexp, just repeat bytes.Index also should work.

```go
if i := bytes.IndexByte(data, '\n'); i >= 0 {
	return i + 1, dropCR(data[0:i]), nil
}
if i := bytes.IndexByte(data, '\r'); i >= 0 {
	return i + 1, dropCR(data[0:i]), nil
}
```

Reply:
We should consider three cases `\r`, `\n` and `\r\n`. We will split the text and add a new line using fmt.Fprintf. So if it is `\r\n`, we will append twice, which performs differently from the original text.

Reply:
Yes, and I think regexp (finite-state machine) is faster than twice bytes.IndexByte.

Reply:
> if it is `\r\n`, we will append twice, which performs differently from the original text.

`\r\n` should only be returned once since we will first look for `\n` and return immediately.


> Yes, and I think regexp (finite-state machine) is faster than twice bytes.IndexByte.

```go
package rebench

import (
	"bytes"
	"math/rand"
	"regexp"
	"testing"
	"time"
)

var (
	n     = 100
	lines = make([][]byte, n, n)
	d     = []string{"\r", "\n", "\r\n"}
)

func init() {
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < n; i++ {
		lines[i] = []byte(RandStr(rand.Intn(30)) + d[rand.Intn(3)] + RandStr(rand.Intn(10)))
	}
}

func BenchmarkIndexByteTwice(b *testing.B) {
	for i := 0; i < b.N; i++ {
		for _, line := range lines {
			_ = bytes.IndexByte(line, '\n')
			_ = bytes.IndexByte(line, '\r')
		}
	}
}

func BenchmarkIndexAny(b *testing.B) {
	for i := 0; i < b.N; i++ {
		for _, line := range lines {
			_ = bytes.IndexAny(line, "\r\n")
		}
	}
}

var newlineReg = regexp.MustCompile(`\r?\n|\r\n?`)

func BenchmarkRegexFindIndex(b *testing.B) {
	for i := 0; i < b.N; i++ {
		for _, line := range lines {
			_ = newlineReg.FindIndex(line)
		}
	}
}
```

and bench shows that:

```
go test -bench .
goos: windows
goarch: amd64
pkg: rebench
cpu: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
BenchmarkIndexByteTwice-8        1658876               713.9 ns/op
BenchmarkIndexAny-8               570092              2283 ns/op
BenchmarkRegexFindIndex-8          16407             71927 ns/op
PASS
ok      rebench 5.387s
```

Reply:
And I use a RandStr impl copied from internet.

```go
package rebench

import (
	"math/rand"
	"time"
	"unsafe"
)

const letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"

var src = rand.NewSource(time.Now().UnixNano())

const (
	// 6 bits to represent a letter index
	letterIdBits = 6
	// All 1-bits as many as letterIdBits
	letterIdMask = 1<<letterIdBits - 1
	letterIdMax  = 63 / letterIdBits
)

func RandStr(n int) string {
	b := make([]byte, n)
	// A rand.Int63() generates 63 random bits, enough for letterIdMax letters!
	for i, cache, remain := n-1, src.Int63(), letterIdMax; i >= 0; {
		if remain == 0 {
			cache, remain = src.Int63(), letterIdMax
		}
		if idx := int(cache & letterIdMask); idx < len(letters) {
			b[i] = letters[idx]
			i--
		}
		cache >>= letterIdBits
		remain--
	}
	return *(*string)(unsafe.Pointer(&b))
}
```

Reply:
Thanks for your benchmark. I'll change it.

Reply:
It’s too late, maybe we need a rest for our health.

Reply:
I found another question. If the input is "abc\rdefg\n", we will get a bad split. So I modified the code to fit a variety of situations.

Reply:
Since we just need to splite the input chunk, I think just locating the last valid delimiter is fine.

Use `bytes.LastIndexByte()` to find the last `\n` or `\r` when `\n` not exists. Also rename `SpliteLines` to `SpliteChunks`.

Reply:
It seems that the performance of `LastIndexByte` is not as good as that of `IndexByte`, because `IndexByte` is implemented in assembly.

```
BenchmarkIndexByteTwice-8                1335291               916.0 ns/op             0 B/op          0 allocs/op
BenchmarkLastIndexByteTwice-8             304917              3776 ns/op               0 B/op          0 allocs/op
BenchmarkIndexAny-8                       396318              3175 ns/op               0 B/op          0 allocs/op
BenchmarkRegexFindIndex-8                   8172            128160 ns/op            1605 B/op        100 allocs/op
```

I still recommend using `IndexByte` twice. As for splitting `PART1\rPART2\nPART3` into 3 parts, I don't think it's necessary, unless there is something else I don't understand, please let me know.

Reply:
We should not accumulate data, we should process it as soon as possible.

Reply:
But it seems not a problem here.